oidc-api/src/main/java/uk/gov/di/authentication/oidc/entity/SlidingWindowAlgorithm.java

```diff
@@ -0,0 +1,88 @@
+    }
+
+    public SlidingWindowAlgorithm(ClientRateLimitDataService clientRateLimitDataService) {
+        this(clientRateLimitDataService, Clock.systemUTC());
+    }
+
+    public SlidingWindowAlgorithm(
+            ClientRateLimitDataService clientRateLimitDataService, Clock clock) {
+        this.rateLimitDataService = clientRateLimitDataService;
+        this.nowClock = new NowHelper.NowClock(clock);
+    }
+
+    @Override
+    public RateLimitDecision getClientRateLimitDecision(ClientRateLimitConfig rateLimitConfig) {
+        var clientId = rateLimitConfig.clientID();
+        var rateLimit = rateLimitConfig.rateLimit();
+
+        var currentTimestamp = nowClock.now().toInstant().atZone(ZoneOffset.UTC).toLocalDateTime();
+        var currentPeriod = getTimeToMinPrecision(currentTimestamp);
+        var previousTimestamp = currentTimestamp.minusMinutes(1);
+        var previousPeriod = getTimeToMinPrecision(previousTimestamp);
+
+        long currentCount =
+                rateLimitDataService
+                        .getData(clientId, currentPeriod)
+                        .map(ClientRateLimitData::getRequestCount)
+                        .orElse(0L);
+        long previousCount =
+                rateLimitDataService
+                        .getData(clientId, previousPeriod)
+                        .map(ClientRateLimitData::getRequestCount)
+                        .orElse(0L);
+
+        // Calculate how many seconds into the current period we are in
+        var secondsFromCurrentPeriodInWindow =
+                currentTimestamp.toEpochSecond(ZoneOffset.UTC)
+                        - currentPeriod.toEpochSecond(ZoneOffset.UTC);
+
+        // Calculate how many seconds from the previous period should be included in this window
+        // e.g 20 seconds into current period means the rest of the 60-second window sits in the
+        // last 40 seconds of the previous period
+        var secondsFromPreviousPeriodInWindow =
+                PERIOD_LENGTH_IN_SECONDS - secondsFromCurrentPeriodInWindow;
+
+        // Scale previous period count by the ratio of seconds from the previous period that are in
+        // the
+        // current window, to seconds in the current window
+        // e.g If previous period has 10 requests, and the window is 30 seconds into the current
+        // period, then we would scale 10 by 1/2 (30/60), to give us 5 requests
+        var previousCountInWindow =
+                previousCount
+                        * ((double) secondsFromPreviousPeriodInWindow / PERIOD_LENGTH_IN_SECONDS);
+        if (previousCountInWindow + currentCount > rateLimit) {
+            LOG.warn(
+                    "Client with ID {} has exceeded rate limit. Current count: {}. Limit {}",
+                    rateLimitConfig.clientID(),
+                    (int) previousCountInWindow + currentCount,
+                    rateLimit);
+            return RateLimitDecision.OVER_LIMIT_RETURN_TO_RP;
```

Comment:
interesting that we are specifying 'return to rp' at this stage. Might it be better if the action taken was determined outside of this class? i.e. the algorithm should just report back that the rate limit has been exceeded or not?